File: packages/firebase_analytics/firebase_analytics/android/src/main/java/io/flutter/plugins/firebaseanalytics/FirebaseAnalyticsPlugin.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebaseanalytics;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.os.Parcelable;
import androidx.annotation.NonNull;
import com.google.android.gms.tasks.Task;
import com.google.firebase.FirebaseApp;
import com.google.firebase.analytics.FirebaseAnalytics;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugins.firebase.core.FlutterFirebasePlugin;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/** Flutter plugin for Firebase Analytics. */
public class FirebaseAnalyticsPlugin
    implements FlutterFirebasePlugin, MethodCallHandler, FlutterPlugin {
  private FirebaseAnalytics firebaseAnalytics;
  private MethodChannel methodChannel;
  // Only set registrar for v1 embedder.
  private PluginRegistry.Registrar registrar;
  // Only set activity for v2 embedder. Always access activity from getActivity() method.
  private Activity activity;

  public static void registerWith(PluginRegistry.Registrar registrar) {
    FirebaseAnalyticsPlugin instance = new FirebaseAnalyticsPlugin();
    instance.registrar = registrar;
    instance.onAttachedToEngine(registrar.context(), registrar.messenger());
  }

  private static Bundle createBundleFromMap(Map<String, Object> map) {
    if (map == null) {
      return null;
    }

    Bundle bundle = new Bundle();
    for (Map.Entry<String, Object> jsonParam : map.entrySet()) {
      final Object value = jsonParam.getValue();
      final String key = jsonParam.getKey();
      if (value instanceof String) {
        bundle.putString(key, (String) value);
      } else if (value instanceof Integer) {
        bundle.putInt(key, (Integer) value);
      } else if (value instanceof Long) {
        bundle.putLong(key, (Long) value);
      } else if (value instanceof Double) {
        bundle.putDouble(key, (Double) value);
      } else if (value instanceof Boolean) {
        bundle.putBoolean(key, (Boolean) value);
      } else if (value == null) {
        bundle.putString(key, null);
      } else if (value instanceof Iterable<?>) {
        ArrayList<Parcelable> list = new ArrayList<Parcelable>();

        for (Object item : (Iterable<?>) value) {
          if (item instanceof Map) {
            list.add(createBundleFromMap((Map<String, Object>) item));
          } else {
            throw new IllegalArgumentException(
                "Unsupported value type: "
                    + item.getClass().getCanonicalName()
                    + " in list at key "
                    + key);
          }
        }

        bundle.putParcelableArrayList(key, list);
      } else if (value instanceof Map<?, ?>) {
        bundle.putParcelable(key, createBundleFromMap((Map<String, Object>) value));
      } else {
        throw new IllegalArgumentException(
            "Unsupported value type: " + value.getClass().getCanonicalName());
      }
    }
    return bundle;
  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
    onAttachedToEngine(binding.getApplicationContext(), binding.getBinaryMessenger());
  }

  private void onAttachedToEngine(Context applicationContext, BinaryMessenger binaryMessenger) {
    FirebaseApp.initializeApp(applicationContext);
    firebaseAnalytics = FirebaseAnalytics.getInstance(applicationContext);
    methodChannel = new MethodChannel(binaryMessenger, "plugins.flutter.io/firebase_analytics");
    methodChannel.setMethodCallHandler(this);
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    firebaseAnalytics = null;
    methodChannel = null;
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    switch (call.method) {
      case "Analytics#logEvent":
        handleLogEvent(call, result);
        break;
      case "Analytics#setUserId":
        handleSetUserId(call, result);
        break;
      case "Analytics#setCurrentScreen":
        handleSetCurrentScreen(call, result);
        break;
      case "Analytics#setAnalyticsCollectionEnabled":
        handleSetAnalyticsCollectionEnabled(call, result);
        break;
      case "Analytics#setSessionTimeoutDuration":
        handleSetSessionTimeoutDuration(call, result);
        break;
      case "Analytics#setUserProperty":
        handleSetUserProperty(call, result);
        break;
      case "Analytics#resetAnalyticsData":
        handleResetAnalyticsData(result);
        break;
      case "Analytics#setConsent":
        setConsent(call, result);
        break;
      case "Analytics#setDefaultEventParameters":
        setDefaultEventParameters(call, result);
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  private void handleLogEvent(MethodCall call, Result result) {
    final String eventName = call.argument("name");
    final Map<String, Object> map = call.argument("parameters");
    final Bundle parameterBundle = createBundleFromMap(map);
    firebaseAnalytics.logEvent(eventName, parameterBundle);
    result.success(null);
  }

  private void handleSetUserId(MethodCall call, Result result) {
    final String id = (String) call.arguments;
    firebaseAnalytics.setUserId(id);
    result.success(null);
  }

  private void handleSetCurrentScreen(MethodCall call, Result result) {
    final String screenName = call.argument("screenName");
    final String screenClassOverride = call.argument("screenClassOverride");
    Bundle parameterBundle = new Bundle();
    parameterBundle.putString(FirebaseAnalytics.Param.SCREEN_NAME, screenName);
    parameterBundle.putString(FirebaseAnalytics.Param.SCREEN_CLASS, screenClassOverride);
    firebaseAnalytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW, parameterBundle);
    result.success(null);
  }

  private void handleSetAnalyticsCollectionEnabled(MethodCall call, Result result) {
    final Boolean enabled = call.arguments();
    firebaseAnalytics.setAnalyticsCollectionEnabled(enabled);
    result.success(null);
  }

  private void handleSetSessionTimeoutDuration(MethodCall call, Result result) {
    final Integer milliseconds = call.arguments();
    firebaseAnalytics.setSessionTimeoutDuration(milliseconds);
    result.success(null);
  }

  private void handleSetUserProperty(MethodCall call, Result result) {
    final String name = call.argument("name");
    final String value = call.argument("value");

    firebaseAnalytics.setUserProperty(name, value);
    result.success(null);
  }

  private void handleResetAnalyticsData(Result result) {
    firebaseAnalytics.resetAnalyticsData();
    result.success(null);
  }

  private void setConsent(MethodCall call, Result result) {
    final String adStorage = call.argument("adStorage");
    final String analyticsStorage = call.argument("analyticsStorage");
    HashMap<FirebaseAnalytics.ConsentType, FirebaseAnalytics.ConsentStatus> parameters =
        new HashMap<>();

    if (adStorage != null) {
      FirebaseAnalytics.ConsentStatus adStorageStatus =
          adStorage.equals(io.flutter.plugins.firebaseanalytics.Constants.DENIED)
              ? FirebaseAnalytics.ConsentStatus.DENIED
              : FirebaseAnalytics.ConsentStatus.GRANTED;

      parameters.put(FirebaseAnalytics.ConsentType.AD_STORAGE, adStorageStatus);
    }

    if (analyticsStorage != null) {
      FirebaseAnalytics.ConsentStatus analyticsStorageStatus =
          analyticsStorage.equals(io.flutter.plugins.firebaseanalytics.Constants.DENIED)
              ? FirebaseAnalytics.ConsentStatus.DENIED
              : FirebaseAnalytics.ConsentStatus.GRANTED;

      parameters.put(FirebaseAnalytics.ConsentType.ANALYTICS_STORAGE, analyticsStorageStatus);
    }

    firebaseAnalytics.setConsent(parameters);
    result.success(null);
  }

  private void setDefaultEventParameters(MethodCall call, Result result) {
    HashMap<String, Object> parameters = call.arguments();

    firebaseAnalytics.setDefaultEventParameters(createBundleFromMap(parameters));
    result.success(null);
  }

  @Override
  public Task<Map<String, Object>> getPluginConstantsForFirebaseApp(FirebaseApp firebaseApp) {
    return null;
  }

  @Override
  public Task<Void> didReinitializeFirebaseCore() {
    return null;
  }
}

File: packages/firebase_analytics/firebase_analytics/ios/firebase_analytics/Sources/firebase_analytics/FirebaseAnalyticsPlugin.swift
// Copyright 2025 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import FirebaseAnalytics
import StoreKit

#if canImport(FlutterMacOS)
  import FlutterMacOS
#else
  import Flutter
  import UIKit
#endif

#if canImport(firebase_core)
  import firebase_core
#else
  import firebase_core_shared
#endif

let kFLTFirebaseAnalyticsName = "name"
let kFLTFirebaseAnalyticsValue = "value"
let kFLTFirebaseAnalyticsEnabled = "enabled"
let kFLTFirebaseAnalyticsEventName = "eventName"
let kFLTFirebaseAnalyticsParameters = "parameters"
let kFLTFirebaseAnalyticsAdStorageConsentGranted = "adStorageConsentGranted"
let kFLTFirebaseAnalyticsStorageConsentGranted = "analyticsStorageConsentGranted"
let kFLTFirebaseAdPersonalizationSignalsConsentGranted = "adPersonalizationSignalsConsentGranted"
let kFLTFirebaseAdUserDataConsentGranted = "adUserDataConsentGranted"
let kFLTFirebaseAnalyticsUserId = "userId"

// swift-format-ignore: AlwaysUseLowerCamelCase
let FLTFirebaseAnalyticsChannelName = "plugins.flutter.io/firebase_analytics"

/// Integration-test channel. Used by the example e2e suite to drive campaign
/// URL / universal-link handling against the live Analytics SDK.
// swift-format-ignore: AlwaysUseLowerCamelCase
let FLTFirebaseAnalyticsTestChannelName = "plugins.flutter.io/firebase_analytics_test"

extension FlutterError: Error {}

public class FirebaseAnalyticsPlugin: NSObject, FLTFirebasePluginProtocol, FlutterPlugin,
  FirebaseAnalyticsHostApi
{
  #if os(iOS)
    fileprivate var testChannel: FlutterMethodChannel?
    fileprivate static var sceneDelegateRegistered = false
    fileprivate static var lastHandledCampaignUrl: String?
  #endif

  public static func register(with registrar: any FlutterPluginRegistrar) {
    let binaryMessenger: FlutterBinaryMessenger

    #if os(macOS)
      binaryMessenger = registrar.messenger
    #elseif os(iOS)
      binaryMessenger = registrar.messenger()
    #endif

    let instance = FirebaseAnalyticsPlugin()
    FirebaseAnalyticsHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: instance)
    #if os(iOS)
      registerSceneDelegateIfAvailable(registrar, instance: instance)
      instance.registerCampaignTestChannel(registrar)
    #endif
  }

  func logEvent(event: [String: Any?], completion: @escaping (Result<Void, any Error>) -> Void) {
    guard let eventName = event[kFLTFirebaseAnalyticsEventName] as? String else {
      completion(.success(()))
      return
    }
    let parameters = event[kFLTFirebaseAnalyticsParameters] as? [String: Any]
    Analytics.logEvent(eventName, parameters: parameters)
    completion(.success(()))
  }

  func setUserId(userId: String?, completion: @escaping (Result<Void, any Error>) -> Void) {
    Analytics.setUserID(userId)
    completion(.success(()))
  }

  func setUserProperty(
    name: String, value: String?,
    completion: @escaping (Result<Void, any Error>) -> Void
  ) {
    Analytics.setUserProperty(value, forName: name)
    completion(.success(()))
  }

  func setAnalyticsCollectionEnabled(
    enabled: Bool,
    completion: @escaping (Result<Void, any Error>) -> Void
  ) {
    Analytics.setAnalyticsCollectionEnabled(enabled)
    completion(.success(()))
  }

  func resetAnalyticsData(completion: @escaping (Result<Void, any Error>) -> Void) {
    Analytics.resetAnalyticsData()
    completion(.success(()))
  }

  func setSessionTimeoutDuration(
    timeout: Int64,
    completion: @escaping (Result<Void, any Error>) -> Void
  ) {
    Analytics.setSessionTimeoutInterval(TimeInterval(timeout))
    completion(.success(()))
  }

  func setConsent(
    consent: [String: Bool?],
    completion: @escaping (Result<Void, any Error>) -> Void
  ) {
    var parameters: [ConsentType: ConsentStatus] = [:]
    if let adStorage = consent[kFLTFirebaseAnalyticsAdStorageConsentGranted] as? Bool {
      parameters[.adStorage] = adStorage ? .granted : .denied
    }
    if let analyticsStorage = consent[kFLTFirebaseAnalyticsStorageConsentGranted] as? Bool {
      parameters[.analyticsStorage] = analyticsStorage ? .granted : .denied
    }
    if let adPersonalization =
      consent[kFLTFirebaseAdPersonalizationSignalsConsentGranted] as? Bool
    {
      parameters[.adPersonalization] = adPersonalization ? .granted : .denied
    }
    if let adUserData = consent[kFLTFirebaseAdUserDataConsentGranted] as? Bool {
      parameters[.adUserData] = adUserData ? .granted : .denied
    }
    Analytics.setConsent(parameters)
    completion(.success(()))
  }

  func setDefaultEventParameters(
    parameters: [String: Any?]?,
    completion: @escaping (Result<Void, any Error>) -> Void
  ) {
    Analytics.setDefaultEventParameters(parameters)
    completion(.success(()))
  }

  func getAppInstanceId(completion: @escaping (Result<String?, any Error>) -> Void) {
    let instanceID = Analytics.appInstanceID()
    completion(.success(instanceID))
  }

  func getSessionId(completion: @escaping (Result<Int64?, any Error>) -> Void) {
    Analytics.sessionID { sessionID, error in
      if let error {
        completion(.failure(error))
      } else {
        completion(.success(sessionID))
      }
    }
  }

  func initiateOnDeviceConversionMeasurement(
    arguments: [String: String?],
    completion:
      @escaping (Result<Void, any Error>)
      -> Void
  ) {
    if let emailAddress = arguments["emailAddress"] as? String {
      Analytics.initiateOnDeviceConversionMeasurement(emailAddress: emailAddress)
    }
    if let phoneNumber = arguments["phoneNumber"] as? String {
      Analytics.initiateOnDeviceConversionMeasurement(phoneNumber: phoneNumber)
    }
    if let hashedEmailAddress = arguments["hashedEmailAddress"] as? String,
      let data = hexStringToData(hashedEmailAddress)
    {
      Analytics.initiateOnDeviceConversionMeasurement(hashedEmailAddress: data)
    }
    if let hashedPhoneNumber = arguments["hashedPhoneNumber"] as? String,
      let data = hexStringToData(hashedPhoneNumber)
    {
      Analytics.initiateOnDeviceConversionMeasurement(hashedPhoneNumber: data)
    }
    completion(.success(()))
  }

  func logTransaction(
    transactionId: String,
    completion: @escaping (Result<Void, any Error>) -> Void
  ) {
    #if os(macOS)
      if #available(macOS 12.0, *) {
        logTransactionWithStoreKit(transactionId: transactionId, completion: completion)
      } else {
        completion(
          .failure(
            FlutterError(
              code: "firebase_analytics",
              message: "logTransaction() is only supported on macOS 12.0 or newer",
              details: nil
            )
          )
        )
      }
    #else
      if #available(iOS 15.0, *) {
        logTransactionWithStoreKit(transactionId: transactionId, completion: completion)
      } else {
        completion(
          .failure(
            FlutterError(
              code: "firebase_analytics",
              message: "logTransaction() is only supported on iOS 15.0 or newer",
              details: nil
            )
          )
        )
      }
    #endif
  }

  #if os(macOS)
    @available(macOS 12.0, *)
  #else
    @available(iOS 15.0, *)
  #endif
  private func logTransactionWithStoreKit(
    transactionId: String,
    completion: @escaping (Result<Void, any Error>) -> Void
  ) {
    Task {
      do {
        guard let id = UInt64(transactionId) else {
          completion(
            .failure(
              FlutterError(
                code: "firebase_analytics",
                message: "Invalid transactionId",
                details: nil
              )
            )
          )
          return
        }

        var foundTransaction: Transaction?
        for await result in Transaction.all {
          switch result {
          case .verified(let transaction):
            if transaction.id == id {
              foundTransaction = transaction
              break
            }
          case .unverified:
            continue
          }
        }

        guard let transaction = foundTransaction else {
          completion(
            .failure(
              FlutterError(
                code: "firebase_analytics",
                message: "Transaction not found",
                details: nil
              )
            )
          )
          return
        }

        Analytics.logTransaction(transaction)
        completion(.success(()))
      } catch {
        completion(.failure(error))
      }
    }
  }

  private func hexStringToData(_ hexString: String) -> Data? {
    let length = hexString.count
    guard length % 2 == 0 else { return nil }

    var data = Data(capacity: length / 2)
    var index = hexString.startIndex

    for _ in 0..<(length / 2) {
      let nextIndex = hexString.index(index, offsetBy: 2)
      guard let byte = UInt8(hexString[index..<nextIndex], radix: 16) else {
        return nil
      }
      data.append(byte)
      index = nextIndex
    }

    return data
  }

  public func didReinitializeFirebaseCore(_ completion: @escaping () -> Void) {
    completion()
  }

  public func pluginConstants(for firebaseApp: FirebaseApp) -> [AnyHashable: Any] {
    [:]
  }

  public func firebaseLibraryName() -> String {
    "flutter-fire-analytics"
  }

  public func firebaseLibraryVersion() -> String {
    versionNumber
  }

  public func flutterChannelName() -> String {
    FLTFirebaseAnalyticsChannelName
  }
}

#if os(iOS)
  /// Registers for UIScene callbacks when the host Flutter SDK supports it.
  ///
  /// `addSceneDelegate(_:)` and `FlutterSceneLifeCycleDelegate` were added in
  /// Flutter 3.38. Calling via `responds(to:)` keeps this plugin compiling on
  /// older Flutter versions; the methods below are unused until registration
  /// succeeds.
  private func registerSceneDelegateIfAvailable(
    _ registrar: FlutterPluginRegistrar,
    instance: FirebaseAnalyticsPlugin
  ) {
    let selector = NSSelectorFromString("addSceneDelegate:")
    guard (registrar as AnyObject).responds(to: selector) else {
      return
    }
    _ = (registrar as AnyObject).perform(selector, with: instance)
    FirebaseAnalyticsPlugin.sceneDelegateRegistered = true
  }

  /// Forwards UIScene URL and universal-link opens to Analytics.
  ///
  /// The native SDK only swizzles `UIApplicationDelegate`. Under UIScene, UIKit
  /// delivers those opens here instead, so campaign attribution (`firebase_campaign`)
  /// is lost unless we call `handleOpen` / `handleUserActivity` ourselves.
  ///
  /// Always returns `false` so later plugins (Auth, Messaging, etc.) still receive
  /// the same scene events. Selectors match `FlutterSceneLifeCycleDelegate`
  /// without conforming to that protocol, which is absent from older Flutter SDKs.
  extension FirebaseAnalyticsPlugin {
    @objc(scene:willConnectToSession:options:)
    public func scene(
      _: UIScene,
      willConnectTo _: UISceneSession,
      options connectionOptions: UIScene.ConnectionOptions?
    ) -> Bool {
      handleCampaign(urlContexts: connectionOptions?.urlContexts)
      handleCampaign(userActivities: connectionOptions?.userActivities)
      return false
    }

    @objc(scene:openURLContexts:)
    public func scene(
      _: UIScene,
      openURLContexts urlContexts: Set<UIOpenURLContext>
    ) -> Bool {
      handleCampaign(urlContexts: urlContexts)
      return false
    }

    @objc(scene:continueUserActivity:)
    public func scene(_: UIScene, continue userActivity: NSUserActivity) -> Bool {
      handleCampaign(userActivity: userActivity)
      return false
    }

    fileprivate func handleCampaign(url: URL) {
      Self.lastHandledCampaignUrl = url.absoluteString
      Analytics.handleOpen(url)
    }

    fileprivate func handleCampaign(userActivity: NSUserActivity) {
      if let url = userActivity.webpageURL {
        Self.lastHandledCampaignUrl = url.absoluteString
      }
      Analytics.handleUserActivity(userActivity)
    }

    private func handleCampaign(urlContexts: Set<UIOpenURLContext>?) {
      guard let urlContexts else { return }
      for context in urlContexts {
        handleCampaign(url: context.url)
      }
    }

    private func handleCampaign(userActivities: Set<NSUserActivity>?) {
      guard let userActivities else { return }
      for activity in userActivities {
        handleCampaign(userActivity: activity)
      }
    }

    fileprivate func registerCampaignTestChannel(_ registrar: FlutterPluginRegistrar) {
      let channel = FlutterMethodChannel(
        name: FLTFirebaseAnalyticsTestChannelName,
        binaryMessenger: registrar.messenger()
      )
      testChannel = channel
      channel.setMethodCallHandler { [weak self] call, result in
        guard let self else {
          result(
            FlutterError(
              code: "unavailable",
              message: "Plugin deallocated",
              details: nil
            )
          )
          return
        }
        self.handleTestChannel(call, result: result)
      }
    }

    private func handleTestChannel(
      _ call: FlutterMethodCall,
      result: @escaping FlutterResult
    ) {
      switch call.method {
      case "isSceneDelegateRegistered":
        result(Self.sceneDelegateRegistered)
      case "getLastHandledCampaignUrl":
        result(Self.lastHandledCampaignUrl)
      case "handleCampaignUrl":
        guard let urlString = call.arguments as? String, let url = URL(string: urlString)
        else {
          result(
            FlutterError(
              code: "invalid-argument",
              message: "Expected a URL string",
              details: nil
            )
          )
          return
        }
        handleCampaign(url: url)
        result(nil)
      case "handleCampaignUserActivity":
        guard let urlString = call.arguments as? String, let url = URL(string: urlString)
        else {
          result(
            FlutterError(
              code: "invalid-argument",
              message: "Expected a URL string",
              details: nil
            )
          )
          return
        }
        guard let scene = UIApplication.shared.connectedScenes.first else {
          result(
            FlutterError(
              code: "unavailable",
              message: "No connected UIScene",
              details: nil
            )
          )
          return
        }
        let activity = NSUserActivity(activityType: NSUserActivityTypeBrowsingWeb)
        activity.webpageURL = url
        _ = self.scene(scene, continue: activity)
        result(Self.lastHandledCampaignUrl)
      case "openUrl":
        guard let urlString = call.arguments as? String, let url = URL(string: urlString)
        else {
          result(
            FlutterError(
              code: "invalid-argument",
              message: "Expected a URL string",
              details: nil
            )
          )
          return
        }
        Self.lastHandledCampaignUrl = nil
        UIApplication.shared.open(url, options: [:]) { success in
          result(success)
        }
      default:
        result(FlutterMethodNotImplemented)
      }
    }
  }
#endif
